File: problem_set_3/longest_increasing_subsequence.go
package main

import "fmt"
func lengthOfLIS(nums []int) int {
	result := []int{nums[0]}
	for _, num := range nums[1:] {
		if num > result[len(result)-1] {
			result = append(result, num)
		} else {
			for i, r := range result {
				if num <= r {
					result[i] = num
					break
				}
			}
		}
	}
	return len(result)
}
func main() {
	test_cases := [][]int{
		{10, 9, 2, 5, 3, 7, 101, 18},
		{0, 1, 0, 3, 2, 3},
		{7, 7, 7, 7, 7, 7, 7},
		{4, 10, 4, 3, 8, 9},
		{2, 2},
		{1, 3, 6, 7, 9, 4, 10, 5, 6},
	}
	for _, test_case := range test_cases {
		fmt.Println(lengthOfLIS(test_case))
	}
}

File: problem_set_3/longest_increasing_subsequence.py
def length_of_lis(nums):
    result = [nums[0]]
    for num in nums[1:]:
        if num > result[-1]:
            result.append(num)
        else:
            i = 0
            while i < len(result) and result[i] < num:
                i += 1
            result[i] = num
    return len(result)

if __name__ == '__main__':
    test_cases = [
        [10, 9, 2, 5, 3, 7, 101, 18],  # 4
        [0, 1, 0, 3, 2, 3],  # 4
        [7, 7, 7, 7, 7, 7, 7],  # 1
        [4, 10, 4, 3, 8, 9],  # 3
        [2, 2],  # 1
        [1, 3, 6, 7, 9, 4, 10, 5, 6],  # 6
    ]
    for test in test_cases:
        print(length_of_lis(test))
